internal/pkg/deploy/cloudformation/project.go

```diff
@@ -221,6 +222,79 @@ func (cf CloudFormation) AddEnvToProject(project *archer.Project, env *archer.En
+	concrete, ok := client.(*cloudformation.CloudFormation)
+	if !ok {
+		return "", errors.New("failed to retrieve the region")
+	}
+	return *concrete.Client.Config.Region, nil
+}
+
+// AddPipelineResourcesToProject adds resources needed to support a pipeline
+// to a project. It conditionally creates resources needed for the pipeline
+// in a AWS region in case there's no existing environment
+// (which would have provisioned these resources) for the project
+func (cf CloudFormation) AddPipelineResourcesToProject(project *archer.Project, pipelineName string) error {
+	projectConfig := stack.NewProjectStackConfig(&deploy.CreateProjectInput{
+		Project:   project.Name,
+		AccountID: project.AccountID,
+	}, cf.box)
+	previouslyDeployedConfig, err := cf.getLastDeployedProjectConfig(projectConfig)
+	if err != nil {
+		return fmt.Errorf("getting previous deployed stackset %w", err)
+	}
+
+	// if the the tool's account Id and region used by ssm store match one
+	// of the existing environment, then it implies that the customer have
+	// deployed an environment into the accountID + region combo that we are
+	// going to create the pipeline in, so there's no need to provision new
+	// resources.
+
+	// First, check whether we need to update the stackset. Since it's possible
+	// that the pipeline is being created in a accountId+region combo where
+	// there's no existing archer environment.
+	accountList := []string{}
+	shouldAddNewAccountID := true
+	for _, accountID := range previouslyDeployedConfig.Accounts {
+		accountList = append(accountList, accountID)
+		if accountID == project.AccountID {
+			shouldAddNewAccountID = false
+		}
+	}
+
+	if shouldAddNewAccountID {
+		// only update the stackset when a new account is added
+		accountList = append(accountList, project.AccountID)
+		newDeploymentConfig := stack.ProjectResourcesConfig{
+			Version:  previouslyDeployedConfig.Version + 1,
+			Apps:     previouslyDeployedConfig.Apps,
+			Accounts: accountList,
+			Project:  projectConfig.Project,
+		}
+
+		if err := cf.deployProjectConfig(projectConfig, &newDeploymentConfig); err != nil {
+			return fmt.Errorf("failed to update resources needed to support the pipeline, project: %s, pipeline: %s, error: %w",
+				project.Name, pipelineName, err)
+		}
+	}
+
+	region, err := getRegionFromClient(cf.client)
+	if err != nil {
+		return fmt.Errorf("failed to provision resources needed to support the pipeline, project: %s, pipeline: %s, error: %w",
+			project.Name, pipelineName, err)
+	}
+
+	// conditionally create a new stack instance in a region for a project
+	// if there's no existing stack instance. Have to do this regardless of
+	// whether the stackset has been updated above because we may need to
+	// provision these resources in a new region in an exist account.
+	if err := cf.addNewProjectStackInstances(projectConfig, region); err != nil {
+		return fmt.Errorf("failed to add stack instance for pipeline%s: %w", pipelineName, err)
+	}
```

Comment:
I think the concern that I have is that this will create ECR repositories in this region even though a user did not create an environment in this region. 
Would it be horrible to create the kms key and s3 bucket in the CodePipeline's cfn template?

Reply:
I tried that and the pipeline code because quite messy because it needs to check whether there's an existing environment in that region before creating any resources

Reply:
Yeah I'm okay with addressing it in a separate PR, would you mind capturing it in a GitHub issue?